### PR TITLE
🏗️ Update instance schema in the hub

### PR DESCRIPTION
**Pull request by fredericenard**
Here is the relevant PR to create the table in the hub:

- https://github.com/laminlabs/laminapp-ui/pull/679

### CODE REVIEW

**Comment by fredericenard** [comment]
@falexwolf upon instance migration we need to call `schema_metadata.to_json()` and store it in the hub. 

image

**Comment by falexwolf** [comment]
Great!

I have some comments re storing this in the hub repo.

File: lamindb_setup/_schema_metadata.py [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+
+
+def synchronize_schema():
+    return call_with_fallback_auth(_synchronize_schema)
+
+
+def _synchronize_schema(client: Client):
+    schema_metadata_dict = schema_metadata.to_json()
+
+    schema_encoded = json.dumps(schema_metadata_dict, sort_keys=True).encode("utf-8")
```
Comment (falexwolf) [inline lamindb_setup/_schema_metadata.py:37]: Alternatively, consider bcoding the dictionary prior to hashing it, similar to: https://github.com/insitro/redun/blob/367033e87ad17ad8a1bdb09bec472f846ffe4b4e/redun/bcoding.py#L206-L212

File: lamindb_setup/_schema_metadata.py [hunk 2]
```diff
@@ -0,0 +1,472 @@
...
+    schema_hash = hashlib.sha256(schema_encoded).digest()
```
Comment (falexwolf) [inline lamindb_setup/_schema_metadata.py:38]: I'd use MD5 because the resulting hash is shorter.

**Comment by falexwolf** [comment]
@fredericenard, I'm getting this error now:

```
____________________________________________________________________________________________________________ test_synchronize_new_schema ____________________________________________________________________________________________________________

setup_instance = None

    def test_synchronize_new_schema(setup_instance):
>       is_new, schema = synchronize_schema()

tests/hub-local/test_synchronize_schema.py:14: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
lamindb_setup/_schema_metadata.py:33: in synchronize_schema
    return call_with_fallback_auth(_synchronize_schema)
lamindb_setup/core/_hub_client.py:135: in call_with_fallback_auth
    raise e
lamindb_setup/core/_hub_client.py:128: in call_with_fallback_auth
    result = callable(**kwargs, client=client)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

client = <supabase._sync.client.SyncClient object at 0x127f19420>

    def _synchronize_schema(client: Client):
        schema_metadata = SchemaMetadata()
        schema_metadata_dict = schema_metadata.to_json()
    
        schema_encoded = json.dumps(schema_metadata_dict, sort_keys=True).encode("utf-8")
        schema_hash = hashlib.sha256(schema_encoded).digest()
        schema_uuid = UUID(bytes=schema_hash[:16])
    
        schema = _get_schema_by_id(schema_uuid, client)
    
        is_new = schema is None
        if is_new:
            module_set_info = schema_metadata._get_module_set_info()
            module_ids = "-".join(str(module_info["id"]) for module_info in module_set_info)
            schema = (
                client.table("schema")
                .insert(
                    {
                        "id": schema_uuid.hex,
                        "module_ids": module_ids,
                        "module_set_info": module_set_info,
                        "json": schema_metadata_dict,
                    }
                )
                .execute()
                .data[0]
            )
    
        instance_response = (
            client.table("instance")
            .update({"schema_id": schema_uuid.hex})
            .eq("id", settings.instance._id.hex)
            .execute()
        )
        assert (
>           len(instance_response.data) == 1
        ), f"Instance {settings.instance._id.hex} was not properly linked to schema {schema_uuid.hex}"
E       AssertionError: Instance 71a2057efcb55b01823887e5558926db was not properly linked to schema 52dcd0f43a173aec3972705497ad8315

lamindb_setup/_schema_metadata.py:71: AssertionError
-------------------------------------------------------
```

File: tests/hub-local/test_synchronize_schema.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+
+@pytest.fixture
+def setup_instance():
+    ln_setup.init(storage="./testdb", schema="bionty")
+    yield
+    ln_setup.delete("testdb", force=True)
+
+
+def test_synchronize_new_schema(setup_instance):
```
Comment (falexwolf) [inline tests/hub-local/test_synchronize_schema.py:13]: Can you please add a test for the hash/uuid?

**Comment by fredericenard** [comment]
@falexwolf it's ok to merge on my end

**Comment by falexwolf** [comment]
Ok, I'll add it to the migrations deploy command before!

**Pull request merged**